api/src/main/java/org/apache/iceberg/types/Types.java

```diff
@@ -723,6 +723,9 @@ public int hashCode() {
 
   public static class StructType extends NestedType {
     private static final Joiner FIELD_SEP = Joiner.on(", ");
+    private static final int NO_HASHCODE = Integer.MIN_VALUE;
```

Comment:
I was originally going to suggest we follow the same pattern we do in `CharSequenceWrapper` where we have two fields, the hashCode and a boolean `hashIsZero` flag. This way in case the hashCode is actually zero we don't recompute it. 

In the current implementation, we would recompute the hashCode if it's actually Integer.MIN_VALUE but arguably not worth the complexity to handle that for this case.